src/core/Ekubo.sol

```diff
@@ -0,0 +1,372 @@
+            feeOnTransfer,
+            hashMul,
+            hashMod,
+            fills,
+            amountOutMin
+        );
+        bytes memory encodedBuyAmount =
+            _setOperatorAndCall(address(CORE), data, uint32(IEkuboCallbacks.locked.selector), _ekuboLockCallback);
+        // buyAmount = abi.decode(abi.decode(encodedBuyAmount, (bytes)), (uint256));
+        assembly ("memory-safe") {
+            // We can skip all the checks performed by `abi.decode` because we know that this is the
+            // verbatim result from `lockAcquired` and that `lockAcquired` encoded the buy amount
+            // correctly.
+            buyAmount := mload(add(0x60, encodedBuyAmount))
+        }
+    }
+
+    function sellToEkuboVIP(
+        address recipient,
+        bool feeOnTransfer,
+        uint256 hashMul,
+        uint256 hashMod,
+        bytes memory fills,
+        ISignatureTransfer.PermitTransferFrom memory permit,
+        bytes memory sig,
+        uint256 amountOutMin
+    ) internal returns (uint256 buyAmount) {
+        bytes memory data = Encoder.encodeVIP(
+            uint32(IEkuboCore.lock.selector),
+            recipient,
+            feeOnTransfer,
+            hashMul,
+            hashMod,
+            fills,
+            permit,
+            sig,
+            _isForwarded(),
+            amountOutMin
+        );
+        bytes memory encodedBuyAmount =
+            _setOperatorAndCall(address(CORE), data, uint32(IEkuboCallbacks.locked.selector), _ekuboLockCallback);
+        // buyAmount = abi.decode(abi.decode(encodedBuyAmount, (bytes)), (uint256));
+        assembly ("memory-safe") {
+            // We can skip all the checks performed by `abi.decode` because we know that this is the
+            // verbatim result from `lockAcquired` and that `lockAcquired` encoded the buy amount
+            // correctly.
+            buyAmount := mload(add(0x60, encodedBuyAmount))
+        }
+    }
+
+    function _ekuboLockCallback(bytes calldata data) private returns (bytes memory) {
+        // We know that our calldata is well-formed. Therefore, the first slot is ekubo lock id,
+        // second slot is 0x20 and third is the length of the strict ABIEncoded payload
+        assembly ("memory-safe") {
+            data.length := calldataload(add(0x40, data.offset))
+            data.offset := add(0x60, data.offset)
+        }
+        return locked(data);
+    }
+
+    function _ekuboPay(
+        IERC20 sellToken,
+        address payer,
+        uint256 sellAmount,
+        ISignatureTransfer.PermitTransferFrom calldata permit,
+        bool isForwarded,
+        bytes calldata sig
+    ) private returns (uint256) {
+        if (sellToken == ETH_ADDRESS) {
+            // TODO: Check if this is a valid case
+            // No need to call pay, the receive function will be used to pay the debt
+            SafeTransferLib.safeTransferETH(payable(address(CORE)), sellAmount);
+            return sellAmount;
```

Comment:
interesting. so this is part of the difference between UniV4 and Ekubo where payment isn't done using a `sync`/`settle` flow, but instead a callback?

Reply:
yep, they don't have `sync`/`settle`. You need to call `pay` specifying a token to repay and they will invoke `payCallback` to collect the payment. That is for tokens. For native eth you just need to send it directly